aws/s3/s3.go

```diff
@@ -0,0 +1,369 @@
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
+	}
+
+	_, err := s.client.HeadObject(context.TODO(), &input)
+	if err == nil {
+		return true, nil
+	}
+
+	var nf *types.NotFound
+	if errors.As(err, &nf) {
+		return false, nil
+	}
+	return false, err
+}
+
+// List returns a list of object keys that match the provided prefix.
+// It will not return more than the specified max number of keys.
+// Keys are in alphabetical order.
+func (s *S3) List(bucket, prefix string, max int32) ([]string, error) {
+	input := s3.ListObjectsV2Input{
+		Bucket:  aws.String(bucket),
+		Prefix:  aws.String(prefix),
+		MaxKeys: max,
+	}
+
+	out, err := s.client.ListObjectsV2(context.TODO(), &input)
+	if err != nil {
+		return nil, err
+	}
+
+	result := make([]string, 0)
+	for _, o := range out.Contents {
+		result = append(result, *o.Key)
+	}
+	return result, nil
+}
+
+// ListAll returns a list of ALL object keys that match the provided prefix.
+// Keys are in alphabetical order.
+func (s *S3) ListAll(bucket, prefix string) ([]string, error) {
+
+	result := make([]string, 0)
+
+	allFound := false
+	var continuationToken *string
+
+	for !allFound {
+
+		input := s3.ListObjectsV2Input{
+			Bucket:            aws.String(bucket),
+			Prefix:            aws.String(prefix),
+			ContinuationToken: continuationToken,
+		}
+
+		out, err := s.client.ListObjectsV2(context.TODO(), &input)
+		if err != nil {
+			return nil, err
+		}
+		for _, o := range out.Contents {
+			result = append(result, *o.Key)
+		}
+		// When result is not truncated, it means all matching keys have been found.
+		if !out.IsTruncated {
+			allFound = true
+		} else {
+			continuationToken = out.NextContinuationToken
+		}
+	}
+	return result, nil
+}
+
+// Returns whether there is an object in bucket with specified prefix.
+func (s *S3) PrefixExists(bucket, prefix string) (bool, error) {
+	input := s3.ListObjectsV2Input{
+		Bucket:  aws.String(bucket),
+		Prefix:  aws.String(prefix),
+		MaxKeys: 1,
+	}
+	out, err := s.client.ListObjectsV2(context.TODO(), &input)
+	if err != nil {
+		return false, err
+	}
+	if len(out.Contents) > 0 {
+		return true, nil
+	} else {
+		return false, nil
+	}
+}
+
+// ListCommonPrefixes returns a list of ALL common prefixes (no 1000 limit).
+func (s *S3) ListCommonPrefixes(bucket, prefix, delimiter string) ([]types.CommonPrefix, error) {
+
+	result := make([]types.CommonPrefix, 0)
+
+	allFound := false
+	var continuationToken *string
+
+	for !allFound {
+
+		input := s3.ListObjectsV2Input{
+			Bucket:            aws.String(bucket),
+			Prefix:            aws.String(prefix),
+			Delimiter:         aws.String(delimiter),
+			ContinuationToken: continuationToken,
+		}
+
+		out, err := s.client.ListObjectsV2(context.TODO(), &input)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, out.CommonPrefixes...)
+
+		// When result is not truncated, it means all common prefixes have been found.
+		if !out.IsTruncated {
+			allFound = true
```

Comment:
As above, this could just be `return result, nil` and there's no need for the allFound variable or the return at the end (note there are other similar returns when errors are handled).  Also, in which case the trailing `else` is redundant.